### PR TITLE
Added tests for onSendProgress

### PR DESCRIPTION
**Pull request by SpeedReach**
### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### CODE REVIEW

File: dio/test/upload_test.dart [hunk 1]
```diff
@@ -97,6 +97,35 @@ void main() {
...
+      '/put',
+      data: stream,
+      onSendProgress: (a, b) {
+        expect(b, 10);
+        collected.add(a);
+      },
+      options: Options(
+        contentType: Headers.textPlainContentType,
+        headers: {
+          Headers.contentLengthHeader: data.length * 2, // set content-length
```
Comment (AlexV525) [inline dio/test/upload_test.dart:121]: You mean 5 * 2 right? We should use `data.expand` to get the length without magic calculations.

File: dio/test/upload_test.dart [hunk 2]
```diff
@@ -97,6 +97,35 @@ void main() {
...
+        },
+      ),
+    );
+    for (int i = 0; i < data.length; i++) {
+      expect(collected[i], (i + 1) * 2);
```
Comment (AlexV525) [inline dio/test/upload_test.dart:126]: Same as above.

File: dio/test/upload_test.dart [hunk 3]
```diff
@@ -97,6 +97,35 @@ void main() {
...
+      Uint8List.fromList([1, 1]),
+      Uint8List.fromList([2, 4]),
+    ];
+    final stream = Stream.fromIterable(data);
+    final List<int> collected = [];
+    final _ = await dio.put(
+      '/put',
+      data: stream,
+      onSendProgress: (a, b) {
+        expect(b, 10);
```
Comment (AlexV525) [inline dio/test/upload_test.dart:115]: Same as others. Once we modified the bytes, 10 becomes invalid and no one knows what it means.

**Comment by SpeedReach** [comment]
@AlexV525 Thanks for the review! I modified the code above, it allows any data bytes length now.

**Comment by AlexV525** [comment]
I saw tests failed, could you resolve them?

**Comment by SpeedReach** [comment]
Hi @AlexV525  
I found that the reason browser's  tests do not pass is because it has different implementations.
The browser one triggers progressEvent only once , so onSendProgress only has two states, 0% and 100%, this should be related to the [dart:html](https://github.com/dart-lang/html/tree/main) package.
However I changed the tests, and they should pass now.
Let me know if there are any further modifications needed.

**Comment by AlexV525** [comment]
> The browser one triggers progressEvent only once , so onSendProgress only has two states, 0% and 100%, this should be related to this [issue](https://github.com/dart-lang/sdk/issues/43904).

Consider adding the ref as a TODO?

**Comment by SpeedReach** [comment]
I could give it a try but I havn't worked on a sdk before.
Should I open a issue for tracking it?

**Comment by SpeedReach** [comment]
By the way this pull request tests if the onSendProgress is triggered properly, but not testing how frequently it is triggered, since dio isn't responsible for it.

**Comment by AlexV525** [comment]
What I mean is to write a TODO comment like:
```dart
// TODO(anyone): Update the test once https://github.com/dart-lang/sdk/issues/43904 has been resolved.
```

**Review by AlexV525 — APPROVED** [review]
LGTM

**Comment by SpeedReach** [comment]
I don't have the permission to merge it.Could you help me merge?

**Pull request merged**